File: bundles/org.eclipse.rap.rwt/src/org/eclipse/swt/internal/graphics/GCAdapter.java
/*******************************************************************************
 * Copyright (c) 2010 EclipseSource and others. All rights reserved.
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v1.0 which accompanies this distribution,
 * and is available at http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   EclipseSource - initial API and implementation
 ******************************************************************************/
package org.eclipse.swt.internal.graphics;

import java.util.LinkedList;
import java.util.List;

import org.eclipse.swt.graphics.Font;
import org.eclipse.swt.internal.graphics.GCOperation.SetFont;


public final class GCAdapter implements IGCAdapter {

  private final List gcOperations;
  private Font lastFont = null;

  public GCAdapter() {
    gcOperations = new LinkedList();
  }

  public void addGCOperation( final GCOperation operation ) {
    // [if] Filter sequential equals SetFont operations as result of multiple GC
    // instances see bug 323080
    if( operation instanceof SetFont ) {
      Font font = ( ( SetFont )operation ).font;
      if( !font.equals( lastFont ) ) {
        lastFont = font;
        gcOperations.add( operation );
      }
    } else {
      gcOperations.add( operation );
    }
  }

  public GCOperation[] getGCOperations() {
    GCOperation[] result = new GCOperation[ gcOperations.size() ];
    gcOperations.toArray( result );
    return result;
  }

  public void clearGCOperations() {
    lastFont = null;
    gcOperations.clear();
  }
}

File: bundles/org.eclipse.rap.rwt/js/org/eclipse/swt/graphics/GCAdapter.js
/*******************************************************************************
 * Copyright (c) 2011, 2012 EclipseSource and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    EclipseSource - initial API and implementation
 ******************************************************************************/

org.eclipse.rwt.protocol.AdapterRegistry.add( "rwt.GC", {

  factory : function( properties ) {
    var parent = org.eclipse.rwt.protocol.ObjectManager.getObject( properties.parent );
    return new org.eclipse.swt.graphics.GC( parent );
  },

  destructor : function( gc ) {
    gc.dispose();
  },

  properties : [],

  propertyHandler : {},

  listeners : [],

  listenerHandler : {},

  methods : [ "init", "draw" ],

  methodHandler : {
    "init" : function( gc, properties ) {
      gc.init(
        properties.width,
        properties.height,
        properties.font,
        properties.fillStyle,
        properties.strokeStyle
      );
    },
    "draw" : function( gc, properties ) {
      gc.draw( properties.operations );
    }
  }


} );
